### ISSUE TITLE
fix: update card in Stripe on expiration date change

### ISSUE THREAD

**Issue by mikejhale** [impress-org/givewp#4280]
# Bug Report

## User Story

As a donor, I want to be able to use a credit card with a new expiration date so that I can successfully make a donation.

### From Customer: 
We had a donation on Sept 2, 2019 at 10:40 a.m. from a customer who had made a previous donation on April 3, 2019. The customer used the same card for both transactions, however the card expired and was renewed in July 2019. The customer entered the new expiry date in the new transaction, but Stripe processed the transaction with the old expiry date and the payment failed with a code of "Do Not Honor." We experienced this issue previously (See #37692). Stripe support pointed me to this page - https://stripe.com/docs/saving-cards#automatic-card-updates - which says that they work with card networks and automatically attempt to update saved card details whenever a customer receives a new card or replaces an expired card. However, while this is widely supported in the US the international support for this varies from country to country. So they couldn't promise that it would work for Canadian cards.

We are finding between 2 to 5 donors per week who have renewed their cards, but are receiving transaction failed messages because Stripe has not received the new expiry date for their cards.

When a repeat donor makes a new donation in our donation form with a new expiry date on a previously used card

1. Givewp pulls the customer record down from Stripe based on the donor email. This includes all the existing cards/sources stored for the donor/customer
3. Givewp pulls the Payment Method record down from Stripe based on the Payment Method Id received from js.stripe.com in the donation data
5. Givewp compares the fingerprint from that Payment Method with the fingerprints from all the existing sources stored for the donor
7. When Givewp finds a match, it moves forward with the existing source that matched the Payment Method.

The problem is that Stripe is giving the same fingerprint for cards regardless of their expiry date. So, the expiry date needs to be compared, along with the fingerprint, in order to verify a match. You can find the problem code in the [attach_payment_method function in class-give-stripe-customer.php](https://github.com/impress-org/givewp/blob/d96dd346408bf1373a395205c29bbc77b7cabfa5/includes/gateways/stripe/includes/class-give-stripe-customer.php#L434). It is in the foreach loop. It is only looking for fingerprint matches, not fingerprint plus expiry date matches. Thus, when it finds a match, it grabs the data from the existing source and loses the new expiry date.

### From Stripe
Your observations of the card fingerprint behavior is spot on; If it's essentially the same card, but details have changed, then it will carry the same fingerprint.

I'd say that updating card details versus deleting and adding new cards, would be the way to go. Stripe does not check the CVC code; This happens on the bank/card issuer side of a transaction, and Stripe passes information back and forth (passing the CVC to the bank, and responding with a failure should the bank see that the CVC entered wasn't matching).

Now, since the checks are made against card fingerprints, and you can't ever be certain that expiration dates have changed until after the details have been sent to the bank/card issuer, one approach might be to update the payment method details each time. I've honestly not seen this issue arise before, so I'm a bit surprised to see this issue popping up so often.

## Current Behavior

If the donor is using a card that already exists in Stripe, Give is pulling the card information with the old expiry date from Stripe and telling Stripe to use that information to process the charge, rather than updating the card first with the new expiry date the user just input and then processing the charge.

## Expected Behavior

When a credit card expiration has changed, update the card in Stripe. 

## Bug Type

- [ ] This bug describes functionality that once worked as expected in version X.X.X.
- [ ] This bug describes functionality that never worked as expected.
- [ ] I am not sure whether this functionality ever worked as expected.

## Steps to Reproduce

1. Make a donation using Stripe
2. Make another donation using the same card, but with a new Expiration/CVC (require the customer to enter all their CC info)
3. See rejected status of donation

## Possible Solution

When comparing the CC fingerprint, also check the expiration data/cvc and update the card in Stripe if it has changed. 

Testing this may be difficult using test credit cards base don how Stripe saved credit card data. 

## Related

#4108 
https://give.productboard.com/feature-board/879919-givewp-feature-hierarchy/features/3105148/insights

## Acceptance Criteria

- [x] Donation processed when using an existing card in Stripe with a new expiration date. 
- [x] Donation processed when using an existing card doesn't create a new card
- [x] Donation processed with a new card should make the card default
- [x] Donation processed with a card when the customer is the first time donor then make the card default
- [x] Test donation processing with Stripe CC and Stripe legacy Checkout (as it uses token)

**Comment by JasonTheAdams** [comment]
I just spoke with the engineers at Stripe and can confirm that the fingerprint only reflects the card number. So you'll need to check the expiration date, as the customer suggests, as well.

According to the Stripe [PaymentMethod docs](https://stripe.com/docs/api/payment_methods/update) you can update a card's expiration month and date as part of the update method. It seems like that's what Stripe is suggesting, and it also means avoiding having to change PaymentMethod ids on attached Customers and Subscriptions.

So there should be a means of finding the existing card and updating it when the user provides an updated card.

**Comment by Benunc** [comment]
From the customer:
> For some cards, and particularly Canadian cards, the CVC code is changed when a card is renewed, so updating the expiry date in a Payment Method will not be enough. A new card/payment method would need to be created with the new expiry date and CVC code.

**Comment by mehul0810** [comment]
**Update:** I've tried reproducing this issue. But, not exactly I can reproduce the issue. However, I found a couple of things that can be cause behind the issue.
1. The payment methods which are attached are not set as default.
2. Even the same fingerprint payment methods are getting added after the update from sources to payment methods.

I've updated the code and had the fix ready. But, before release, I want to test this with lots of possible scenarios with different payment options provided by Stripe. So, I'll create a PR about this issue on Monday.
